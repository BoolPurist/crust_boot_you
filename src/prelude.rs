use std::io;

pub use crate::constants;
pub use anyhow::Context;

#[cfg(test)]
pub use serde::{Deserialize, Serialize};

pub use derive_getters::Getters;
pub use derive_new::new;
pub type IoResult<T = ()> = Result<T, io::Error>;
pub type AppError = anyhow::Error;
pub type AppResult<T = ()> = Result<T, AppError>;

pub type ReturnToUser = AppResult<String>;
pub type PathResult = AppResult<PathBuf>;
pub type AppIoResult<T = ()> = Result<T, AppIoError>;

pub use crate::app_traits::file_manipulator::FileManipulator;
pub use crate::app_traits::path_provider::PathProvider;
pub use crate::app_traits::path_resolver::PathResolver;
pub use crate::file_management::AppIoError;
pub use crate::ValidTemplateName;
pub use std::path::{Path, PathBuf};

#[cfg(test)]
pub use testing::*;
#[cfg(test)]
pub mod testing {

    pub const TEST_INPUT_FOLDER_NAME: &str = "test_input";
    #[macro_export]
    macro_rules! from_ron_input_file {
        ($input:literal) => {{
            let path = const_format::concatcp!(
                TEST_INPUT_FOLDER_NAME,
                std::path::MAIN_SEPARATOR_STR,
                $input
            );
            ron::from_str(load_str!(path)).expect(concat!($input, "is not in a valid format"))
        }};
    }
    pub use from_ron_input_file;
}
