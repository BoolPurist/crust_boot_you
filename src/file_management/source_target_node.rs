//! TODO: Remove if used
#![allow(dead_code)]

use crate::prelude::*;

use super::{FileKind, NodeEntryMeta};
use std::path::PathBuf;

#[derive(Debug)]
#[cfg_attr(test, derive(Serialize, Deserialize))]
pub struct SourceTargetNode {
    source_node: NodeEntryMeta,
    target: PathBuf,
}

impl SourceTargetNode {
    fn new(source_node: NodeEntryMeta, target: PathBuf) -> Self {
        Self {
            source_node,
            target,
        }
    }

    pub fn source_path(&self) -> &Path {
        self.source_node.source_path()
    }
    pub fn target_path(&self) -> &Path {
        self.target.as_path()
    }
    pub fn node_type(&self) -> FileKind {
        *self.source_node.node_type()
    }

    pub fn opt_many_from_many_sources(
        prefix_source: &Path,
        prefix_target: &Path,
        source_to_map: impl IntoIterator<Item = NodeEntryMeta>,
    ) -> Option<Vec<Self>> {
        source_to_map
            .into_iter()
            .map(|node| {
                let target_path = node
                    .source_path()
                    .strip_prefix(prefix_source)
                    .ok()
                    .map(|to_join_with| prefix_target.join(to_join_with))?;
                Some(Self::new(node, target_path))
            })
            .collect()
    }
}

#[cfg(test)]
mod testing {

    use super::*;

    #[test]
    fn test_some_if_valid_suffix_and_prefix() {
        let source = Path::new("/some/source");
        let target = Path::new("/some/target");

        let to_map: Vec<NodeEntryMeta> = from_ron_input_file!("source_paths_with_same_prefix.ron");

        // Act
        let actual = SourceTargetNode::opt_many_from_many_sources(source, target, to_map).unwrap();

        insta::assert_ron_snapshot!(actual);
    }
    #[test]
    fn test_none_if_one_file_with_wrong_prefix() {
        use crate::prelude::testing::*;
        let source = Path::new("/some/source");
        let target = Path::new("/some/target");

        let to_map: Vec<NodeEntryMeta> = from_ron_input_file!("different_prefixes.ron");

        // Act
        let actual = SourceTargetNode::opt_many_from_many_sources(source, target, to_map);

        assert!(actual.is_none());
    }
}
