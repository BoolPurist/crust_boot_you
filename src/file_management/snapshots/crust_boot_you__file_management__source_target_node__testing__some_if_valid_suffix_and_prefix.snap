---
source: src/file_management/source_target_node.rs
expression: actual
---
[
  SourceTargetNode(
    source_node: NodeEntryMeta(
      node_type: File,
      source_path: "/some/source/a",
    ),
    target: "/some/target/a",
  ),
  SourceTargetNode(
    source_node: NodeEntryMeta(
      node_type: Folder,
      source_path: "/some/source/a",
    ),
    target: "/some/target/a",
  ),
  SourceTargetNode(
    source_node: NodeEntryMeta(
      node_type: File,
      source_path: "/some/source/b",
    ),
    target: "/some/target/b",
  ),
]
